payments/models.py

```diff
@@ -154,3 +171,376 @@ def __str__(self):
+    )
+    product = models.ForeignKey(
+        AdditionalProduct,
+        verbose_name=_("product"),
+        related_name="orders_lines",
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+    )
+    quantity = models.PositiveSmallIntegerField(
+        verbose_name=_("quantity"), default=1, validators=[MinValueValidator(1)],
+    )
+    price = models.DecimalField(
+        verbose_name=_("price"),
+        max_digits=10,
+        decimal_places=2,
+        validators=[MinValueValidator(Decimal("0.01"))],
+        blank=True,
+    )
+    tax_percentage = models.DecimalField(
+        verbose_name=_("tax percentage"),
+        max_digits=5,
+        decimal_places=2,
+        blank=True,
+        validators=[MinValueValidator(Decimal("0.00"))],
+    )
+
+    def clean(self):
+        creating = self._state.adding
+        if not creating:
+            old_instance = OrderLine.objects.get(id=self.id)
+            # If the order is being changed
+            if old_instance.order and self.order != old_instance.order:
+                raise ValidationError(
+                    _("Cannot change the order associated with this order line")
+                )
+
+            # If the product is being changed
+            if old_instance.product != self.product:
+                raise ValidationError(
+                    _("Cannot change the product assigned to this order line")
+                )
+
+    def save(self, *args, **kwargs):
+        self.full_clean()
+
+        creating = self._state.adding
+        if creating and self.product:
+            price = self.product.price_value
+            unit = self.product.price_unit
+
+            if unit == PriceUnits.PERCENTAGE:
+                price = calculate_product_percentage_price(
+                    self.order.price, self.product.price_value
+                )
+
+            if self.product.period == PeriodType.MONTH:
+                price = calculate_product_partial_month_price(
+                    price, self.order.lease.start_date, self.order.lease.end_date
+                )
+            elif (
+                self.product.period == PeriodType.SEASON
+                and not self.order.is_full_season()
+            ):
+                # Calculate the actual price for the amount of days on the period
+                # price = (days_on_lease * product.price_value) / season_days
+                price = calculate_product_partial_season_price(
+                    price, self.order.lease.start_date, self.order.lease.end_date
+                )
+            elif self.product.period == PeriodType.YEAR:
+                price = calculate_product_partial_year_price(
+                    price, self.order.lease.start_date, self.order.lease.end_date
+                )
+
+            if not self.price:
+                self.price = price
+            if not self.tax_percentage:
+                self.tax_percentage = self.product.tax_percentage
```

Comment:
That validation is done for [`Order`s](https://github.com/City-of-Helsinki/berth-reservations/pull/214/files#diff-409ee34417f21f35e34b41dfabf4720cR307-R315)
```python
if (
    self.price
    and self.tax_percentage
    and (self.product or self._product_object_id)
):
    raise ValidationError(
        _("Cannot receive product and plain price/tax_percentage")
    )
```

But not for `OrderLine`. Should we also add it here?

Reply:
ok, I gave it some more thought and I believe now that this is a good rule for both `Order` and `OrderLine` models: allow passing both a product and a price. since if it is possible to override the price after an object is created anyway, it is kinda silly that we do not allow it during `create`. if both were passed, price overrides product's price.

Reply:
Alrighty then, let's aim for a standard behaviour and rely on the front-end/API to only allow either of them at a time.